text/0000-code-literals.md

```diff
@@ -0,0 +1,417 @@
+        from the same inability to be reformatted, and the trick
+        of using an `\` at the end of each line cannot be applied
+        because escape characters are not recognised.
+
+  - The existing string literals introduce extra unwanted whitespace
+    into the literal value. Even if that extra whitespace does not
+    semantically affect the nested code, it results in ugly output
+    if the code is ever logged (such as might happen when logging
+    SQL query executions).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In addition to string literals and raw string literals, a third type
+of string literal exists: code string literals.
+
+```rust
+    let code = ```
+        This is a code string literal
+
+        I can use special characters like "" and \ freely.
+
+            Indentation is preserved *relative* to the indentation level
+            of the terminating triple backticks.
+
+    It is an error for a line to have "negative" indentation (ie. be
+    indented less than the final triple backticks) unless
+    the line is empty.
+        ```;
+```
+
+`rustfmt` will automatically adjust the indentation of the code string
+literal as a whole to match the surrounding context, but will never
+change the relative indentation within such a literal.
+
+Anything directly after the opening backticks is not considered
+part of the string literal. It may be used as a language hint or
+processed by macros (similar to the treatment of doc comments).
+
+```rust
+let sql = ```sql
+    SELECT * FROM table;
+    ```;
+```
+
+Similar to raw string literals, there is no way to escape characters
+within a code string literal. It is expected that procedural macros
+would build upon code string literals to add support for such
+functionality as required.
+
+If it is necessary to include triple backticks within a code string
+literal, more than three backticks may be used to enclose the
+literal, eg.
+
+```rust
+let code = ````
+    ```
+````;
+```
+
+In order to suppress the final newline, the literal may instead be
+closed with `!``` `, eg.
+
+```rust
+let code = ```
+    Text with no final newline
+    !```;
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A code string literal will begin and end with three or more backticks.
+The number of backticks in the terminator must match the number used
+to begin the literal.
+
+The value of the string literal will be determined using the following
+steps:
+
+1.  Measure the whitespace indenting the closing backticks. If a
+    non-whitespace character (other than a single `!`) exists before
+    the closing backticks on the same line, then issue a compiler error.
+2.  Take the lines *between* (but not including) the opening and
+    closing backticks exactly as written.
+3.  Remove exactly the measured whitespace from each line. If this
+    cannot be done, then issue a compiler error.
```

Comment:
It would be helpful to define "measured whitespace". Does this mean the exact sequence of tabs and spaces? What if there's the same amount of indentation, but the order is different?  What if some lines are indented with tabs and others with spaces?

Also, the Guide-level explanation said that entirely empty lines were allowed, but this seems to forbid them.